bsp/Src/ADC.c

```diff
@@ -0,0 +1,217 @@
+    // --- Start ADC Interrupt
+    #ifdef ADC1
+    if (h == hadc1) {
+        adc1_q = q;
+        if (uxQueueSpacesAvailable(*adc1_q) == 0) {
+            return ADC_QUEUE_FULL;
+        }
+        if (HAL_ADC_ConfigChannel(hadc1, &sConfig) != HAL_OK) {
+            return ADC_CHANNEL_CONFIG_FAIL;
+        }
+
+        if (HAL_ADC_Start_IT(hadc1) != HAL_OK) {
+            return ADC_INTERRUPT_FAIL;
+        }
+    }
+    #endif
+    #ifdef ADC2
+    if (h == hadc2) {
+        adc2_q = q;
+        if (uxQueueSpacesAvailable(*adc2_q) == 0) {
+            return ADC_QUEUE_FULL;
+        }
+        if (HAL_ADC_ConfigChannel(hadc2, &sConfig) != HAL_OK) {
+            return ADC_CHANNEL_CONFIG_FAIL;
+        }
+
+        if (HAL_ADC_Start_IT(hadc2) != HAL_OK) {
+            return ADC_INTERRUPT_FAIL;
+        }
+    }
+    #endif
+    #ifdef ADC3
+    if (h == hadc3) {
+        adc3_q = q;
+        if (uxQueueSpacesAvailable(*adc3_q) == 0) {
+            return ADC_QUEUE_FULL;
+        }
+        if (HAL_ADC_ConfigChannel(hadc3, &sConfig) != HAL_OK) {
+            return ADC_CHANNEL_CONFIG_FAIL;
+        }
+
+        if (HAL_ADC_Start_IT(hadc3) != HAL_OK) {
+            return ADC_INTERRUPT_FAIL;
+        }
+    }
+    #endif
+    
+    return ADC_OK; 
+}
+
+
+void HAL_ADC_ConvCpltCallback(ADC_HandleTypeDef *h) {
+    /*
+    Interrupt triggers this callback when the conversion is complete
+    */ 
+    BaseType_t higherPriorityTaskWoken = pdFALSE;
+    int rawVal;
+
+    #ifdef ADC1
+    if (h == hadc1) {
+        rawVal = HAL_ADC_GetValue(hadc1);
+        xQueueSendFromISR(*adc1_q, &rawVal, &higherPriorityTaskWoken);
+
+    } 
+    #endif
+    #ifdef ADC2
+    if (h == hadc2) {
+        rawVal = HAL_ADC_GetValue(hadc2);
+        xQueueSendFromISR(*adc2_q, &rawVal, &higherPriorityTaskWoken);
+
+    } 
+    #endif
+    #ifdef ADC3
+    if (h == hadc3) {
+        rawVal = HAL_ADC_GetValue(hadc3);
+        xQueueSendFromISR(*adc3_q, &rawVal, &higherPriorityTaskWoken);
+        
+    }
+    #endif
+}
+
+void ADC_IRQHandler() {
+    // Arbitrate between ADCs
+    #ifdef ADC1
+    #if defined(STM32L4xx) 
+    if (ADC_FLAG_EOC & ADC1->ISR) {
+        HAL_ADC_IRQHandler(hadc1);
+    }
+    #else
+    if (ADC_FLAG_EOC & ADC1->SR) {
+        HAL_ADC_IRQHandler(hadc1);
+    }
+    #endif
+    #endif
+
+    #ifdef ADC2
+    #if defined(STM32L4xx) 
+    if (ADC_FLAG_EOC & ADC2->ISR) {
+        HAL_ADC_IRQHandler(hadc2);
+    }
+    #else
+    if (ADC_FLAG_EOC & ADC2->SR) {
+        HAL_ADC_IRQHandler(hadc2);
+    }
+    #endif
+    #endif
+
+    #ifdef ADC3
+    #if defined(STM32L4xx) 
+    if (ADC_FLAG_EOC & ADC3->ISR) {
+        HAL_ADC_IRQHandler(hadc3);
+    }
+    #else
+    if (ADC_FLAG_EOC & ADC3->SR) {
+        HAL_ADC_IRQHandler(hadc3);
+    }
+    #endif
+    #endif
+}
+
+void HAL_ADC_ErrorCallback(ADC_HandleTypeDef *h) {
+    // in case of error
+    ADC_ERR_CODE = h->ErrorCode;
+
+    // try again
+    HAL_ADC_Stop_IT(h);
+    HAL_ADC_Start_IT(h);
```

Comment:
The error functionality here is a bit simplistic but fine for now. I'd like to push this off to another ticket.